### PR TITLE
ENH: fft: Add `prev_fast_len` to complement `next_fast_len`

### PR DESCRIPTION
**Pull request by tpl2go**
#### Reference issue
none

#### What does this implement/fix?
This PR adds a new function `prev_fast_len` into the `fft` module.
This function finds the largest composite of FFT radices (2,3,5,7,11) that is less than the target length.
Code, docstring, and unit test are included in this PR

Motivation:
A function `next_fast_len` already exists to help find a fast fft length for zeropadding. 
However in some use cases, we may not want to zeropad a signal but instead discard samples to achieve a fast fft len.

#### Additional information
none

PS: This is my first PR to an open source project. Feedback is appreciated if any part falls short

### CODE REVIEW

**Comment by tpl2go** [comment]
@lucascolley thanks for the mention. I wasn't sure if this would fit in the library but at least I use it within my own work. Would be glad to help clean up this PR if there is interest in it

**Review by peterbell10 — COMMENTED** [review]
Appologies for the extremely late response. This seems like a reasonable counterpart to `next_fast_len` and I can see how this would be valuable to use cases where you can discard part of the signal.

I have a few comments for improving the search algorithm, and also the C++ `pocketfft` files are actually just vendored from https://github.com/mreineck/pocketfft so we should try not to diverge from the main repo. (In fact this should probably be a submodule now, the project wasn't on github when it was first introduced to SciPy).

File: scipy/fft/_pocketfft/pocketfft_hdronly.h [hunk 1]
```diff
@@ -449,6 +449,78 @@ struct util // hack to avoid duplicate symbols
...
+          while(f11753 <= n)
+          {
+            size_t f117532 = f11753;
+            while(f117532 <= n)
+            {
+              if (f117532 > bestfound)
+              {
+                bestfound = f117532;
+              }
+              f117532 *= 2;
```
[the remark this answers is not in the document]
Comment (tpl2go) [reply scipy/fft/_pocketfft/pocketfft_hdronly.h:480]: thanks a lot of the explanation! It's pretty smart! I will refactor the code to include this stair-step search.

File: scipy/fft/_pocketfft/pocketfft_hdronly.h [hunk 2]
```diff
@@ -449,6 +449,78 @@ struct util // hack to avoid duplicate symbols
     return bestfac;
     }
 
+  /* returns the largest composite of 2, 3, 5, 7 and 11 which is <= n */
+  static POCKETFFT_NOINLINE size_t prev_good_size_cmplx(size_t n)
```
[the remark this answers is not in the document]
Comment (tpl2go) [reply scipy/fft/_pocketfft/pocketfft_hdronly.h:453]: sure thing! sorry for the late reply too. Had only the weekend to work on this

**Comment by lucascolley** [comment]
Hey @tpl2go , your upstream PR and my PR to move `pocketfft` to a git submodule are both merged now. Would you like to try to update this PR now? Let me know if you need any help. Then we should be able to finish up the review

**Comment by tpl2go** [comment]
@lucascolley sorry I had busy week. yup I can update this PR! 
But before I proceed I think we need to pull the upstream commits from mreineck/pocketfft into scipy/pocketfft. I just saw that you opened a PR a few hours ago at scipy/pocketfft to do just that. Should I wait for your merge before I update the PR?

**Comment by lucascolley** [comment]
> Should I wait for your merge before I update the PR?

Yep sounds good, we've had to pull in new changes anyway for a separate issue

**Comment by lucascolley** [comment]
The submodule is updated with your commits now

File: scipy/fft/_helper.py [hunk 3]
```diff
@@ -74,6 +74,71 @@ def next_fast_len(target, real=False):
...
+    >>> c = fft.fft(a[:92160]) # discard last 899 samples
+
+    """
+    pass
+
+
+# Directly wrap the c-function prev_good_size but take the docstring etc.,
+# from the prev_fast_len function above
+prev_fast_len = update_wrapper(lru_cache()(_helper.prev_good_size),
+                                prev_fast_len)
```
Comment (lucascolley) [inline scipy/fft/_helper.py:138]:
```suggestion
prev_fast_len = update_wrapper(lru_cache()(_helper.prev_good_size), prev_fast_len)
```

**Comment by tpl2go** [comment]
okie I think this should work! seems like all CI tests passed. 
Anything else that I might have missed?
PS: Thanks for the help @lucascolley

File: scipy/fft/_helper.py [hunk 4]
```diff
@@ -74,6 +74,70 @@ def next_fast_len(target, real=False):
...
+
+    """
+    pass
+
+
+# Directly wrap the c-function prev_good_size but take the docstring etc.,
+# from the prev_fast_len function above
+prev_fast_len = update_wrapper(lru_cache()(_helper.prev_good_size), prev_fast_len)
+prev_fast_len.__wrapped__ = _helper.prev_good_size
+
```
Comment (j-bowhay) [inline scipy/fft/_helper.py:139]: This is needed for the signature to correctly appear in the docs
```suggestion
# from the prev_fast_len function above
_sig_prev_fast_len = inspect.signature(prev_fast_len)
prev_fast_len = update_wrapper(lru_cache()(_helper.prev_good_size), prev_fast_len)
prev_fast_len.__wrapped__ = _helper.prev_good_size
prev_fast_len.__signature__ = _sig_prev_fast_len

```

**Review by lucascolley — APPROVED** [review]
LGTM but I'm not qualified to check the C++ parts - if you could give a final check @peterbell10 that would be great!

**Comment by tylerjereddy** [comment]
There's a request for C++ code review here so I'm bumping the milestone given (somewhat) imminent branching for `1.14.0`. If it gets merged before we branch, just switch the milestone back.

**Review by peterbell10 — APPROVED** [review]

**Comment by lucascolley** [comment]
thanks for the review @peterbell10 !

and congrats on your first contribution to SciPy @tpl2go ! Thanks for all of your patience here 🙏

**Pull request merged**